node/pkg/watchers/sui/watcher.go

```diff
@@ -0,0 +1,355 @@
+package sui
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"net/http"
+	"strings"
+	"time"
+
+	"encoding/base64"
+
+	"golang.org/x/net/websocket"
+
+	"github.com/certusone/wormhole/node/pkg/common"
+	"github.com/certusone/wormhole/node/pkg/p2p"
+	gossipv1 "github.com/certusone/wormhole/node/pkg/proto/gossip/v1"
+	"github.com/certusone/wormhole/node/pkg/readiness"
+	"github.com/certusone/wormhole/node/pkg/supervisor"
+
+	eth_common "github.com/ethereum/go-ethereum/common"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+
+	"github.com/tidwall/gjson"
+	"github.com/wormhole-foundation/wormhole/sdk/vaa"
+	"go.uber.org/zap"
+)
+
+type (
+	// Watcher is responsible for looking over Sui blockchain and reporting new transactions to the wormhole contract
+	Watcher struct {
+		suiRPC     string
+		suiWS      string
+		suiAccount string
+		suiPackage string
+
+		unsafeDevMode bool
+
+		msgChan  chan *common.MessagePublication
+		obsvReqC chan *gossipv1.ObservationRequest
+
+		subId      int64
+		subscribed bool
+	}
+)
+
+var (
+	suiMessagesConfirmed = promauto.NewCounter(
+		prometheus.CounterOpts{
+			Name: "wormhole_sui_observations_confirmed_total",
+			Help: "Total number of verified Sui observations found",
+		})
+	currentSuiHeight = promauto.NewGauge(
+		prometheus.GaugeOpts{
+			Name: "wormhole_sui_current_height",
+			Help: "Current Sui block height",
+		})
+)
+
+// NewWatcher creates a new Sui appid watcher
+func NewWatcher(
+	suiRPC string,
+	suiWS string,
+	suiAccount string,
+	suiPackage string,
+	unsafeDevMode bool,
+	messageEvents chan *common.MessagePublication,
+	obsvReqC chan *gossipv1.ObservationRequest,
+) *Watcher {
+	return &Watcher{
+		suiRPC:        suiRPC,
+		suiWS:         suiWS,
+		suiAccount:    suiAccount,
+		suiPackage:    suiPackage,
+		unsafeDevMode: unsafeDevMode,
+		msgChan:       messageEvents,
+		obsvReqC:      obsvReqC,
+		subId:         0,
+		subscribed:    false,
+	}
+}
+
+func (e *Watcher) inspectBody(logger *zap.Logger, body gjson.Result) error {
+	txDigest := body.Get("txDigest")
```

Comment:
Given this has a static schema I'd recommend using the struct decoder. 

You can turn json into a struct here: https://mholt.github.io/json-to-go/